src/main/java/com/mongodb/spark/sql/connector/schema/RowToBsonDocumentConverter.java

```diff
@@ -165,17 +205,21 @@ public BsonValue toBsonValue(final DataType dataType, final Object data) {
         }
         for (Object obj : listData) {
           if (!(ignoreNulls && Objects.isNull(obj))) {
-            bsonArray.add(toBsonValue(elementType, obj));
+            bsonArray.add(element.apply(obj));
           }
         }
         return bsonArray;
-      } else if (dataType instanceof MapType) {
+      };
+    } else if (dataType instanceof MapType) {
+      return (data) -> {
         DataType keyType = ((MapType) dataType).keyType();
         DataType valueType = ((MapType) dataType).valueType();
         if (!(keyType instanceof StringType)) {
           throw new DataException(
               format("Cannot cast %s into a BsonValue. Invalid key type %s.", data, keyType));
         }
+        Function<Object, BsonValue> value =
```

Comment:
To avoid a performance penalty here, should this be lifted out of the lambda?

Reply:
Done.